File: DFS/1031.Is Graph Bipartite/Solution_BFS.py
from collections import deque


class Solution:
    """
    @param graph: the given undirected graph
    @return:  return true if and only if it is bipartite
    """
    def isBipartite(self, graph):
        # Write your code here
        if not graph or len(graph) == 0:
            return False

        n = len(graph)
        visited = [0 for _ in range(n)]
        color = [0 for _ in range(n)]


        for i in range(n):
            if len(graph[i]) > 0 and visited[i] == 0:
                visited[i] = 1
                color[i] = 1
                queue = deque([i])

                while queue:
                    node = queue.popleft()
                    for neighbor in graph[node]:
                        if visited[neighbor]:
                            if color[node] == color[neighbor]:
                                return False
                        else:
                            visited[neighbor] = 1
                            color[neighbor] = - color[node]
                            queue.append(neighbor)

        return True

File: DFS/1031.Is Graph Bipartite/Solution_BFS.java
public class Solution {
    /**
     * @param graph: the given undirected graph
     * @return:  return true if and only if it is bipartite
     */
    public boolean isBipartite(int[][] graph) {
        // Write your code here
        //BFS
        // 0(not meet), 1(black), 2(white)
        int[] visited = new int[graph.length];

        for (int i = 0; i < graph.length; i++) {
            if (graph[i].length != 0 && visited[i] == 0) {

                visited[i] = 1;
                Queue<Integer> q = new LinkedList<>();
                q.offer(i);

                while(!q.isEmpty()) {
                    int current = q.poll();
                    for (int c: graph[current]) {

                            if (visited[c] == 0) {
                                visited[c] = (visited[current] == 1) ? 2 : 1;
                                q.offer(c);
                            } else {
                                if (visited[c] == visited[current]) {
                                    return false;
                                }
                            }
                    }
                }                        
                
            }
        }

        return true;
    }
}